api/analyzedphenotypes.datafile.api.inc

```diff
@@ -0,0 +1,175 @@
+<?php
+/**
+ * @file
+ *
+ * API: DATA FILE
+ * Manage property and operation pertaining to tsv or txt data file.
+ */
+
+
+/**
+ * DATA SOURCE
+ * Define file extension this module supports.
+ *
+ * @param $set
+ *   A string value requesting a subset of file extensions required in file upload
+ *   and file extension used in data export. Default to null which will return
+ *   all or both subsets.
+ *
+ * @return
+ *   An array, containing file extensions.
+ */
+function ap_define_datafile_extensions($set = null) {
+  $extensions = array(
+    'file_upload'   => array(
+      'tsv'  => 'Tab Separated Values',
+      'txt'  => 'Text File',
+    ),
+    'file_download' => array(
+      'tsv'  => '.TSV - Tab Separated Values',
+      'csv'  => '.CSV - Comman Separated Values',
+      'xlsx' => '.XLSX - Microsoft Excel Spreadsheet File',
+    )
+  );
+
+  return ($set) ? $extensions[$set] : $extensions;
+}
+
+/**
+ * Define file destination path for file upload and trait image.
+ * 
+ * @return 
+ *   Associative array where the key is directory type and value as directory path.
+ */
+function ap_construct_datafile_destinationpath() {
+  $drupal_public = 'public://';
+  $ap_root_dir = 'tripal/analyzedphenotypes/';
+  
+  // Add directory here. 
+  $directory = array(
+    'spreadsheet',
+    'traitpictures'
+  );
+  
+  // Constuct directories.
+  $file_dir = array();
+  
+  foreach($directory as $dir) {
+    $file_dir[$dir] = $drupal_public . $ap_root_dir . $dir; 
+  }
+  
+  return $file_dir;
+}
+
+///////
+
+
+/**
+ * CONTROL/SWITCHBOARD.
+ * Pass a request to a corresponding function.
+ * @TODO: leave out when api reorganize/update is complete and use functions below.
+ */
+function analyzedphenotypes_datafileprop($property = null, $parameter = null) {
+  if (is_null($property)) {
+    // Default.
+    $property = 'default';
+  }
+
+  // Process request.
+  // Default to all projects.
+  $command = array(
+    'default' => array(
+      'func'  => 'ap_define_datafile_extensions',
+      'param' => 'file_upload',
+    ),
+    'download_type' => array(
+      'func'  => 'ap_define_datafile_extensions',
+      'param' => 'file_download',
+    ),
+    'get_contents' => array(
+      'func'  => 'ap_get_datafile_contents',
+      'param' => $parameter['data_file'],
+    ),
+    'get_columns' => array(
+      'func'  => 'ap_get_datafile_columns',
+      'param' => $parameter['data_file'],
+    ),
+    'make_permanent' => array(
+      'func'  => 'ap_make_datafile_permanent',
+      'param' => $parameter['data_file'],
+    ),
+  );
+
+  return call_user_func($command[$property]['func'], $command[$property]['param']);
+}
+
+
+///////
+
+
+/**
+ * FUNCTION:
+ * Read data file and return file data contents.
+ *
+ * @param $datafile
+ *   Drupal file object.
+ *
+ * @return
+ *   File contents.
+ */
+function ap_get_datafile_contents($datafile) {
```

Comment:
Reading a whole file into memory is almost never a good idea when dealing with bioloigcal data. How is this function being used?

I think this function should cease to exist...